### PR TITLE
Add docker build preparations

### PR DESCRIPTION
**Pull request by EdenWuyifan**
To build the docker image, run this at Dockerfile level:
`docker build --build-arg TOKEN='' --build-arg BUILD_OPTION='full' .`

To run the docker image, run:
`docker run --rm -d -p 8899:8888 aaaaaaaaa`

You can then access the jupyter notebook at localhost at the docker port.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ENV HOME=/home/eden
+WORKDIR $HOME
+USER eden
+COPY --chown=eden examples /home/eden/examples
+
+ARG TOKEN=-1
+RUN mkdir -p $HOME/.jupyter/
+RUN if [ $TOKEN!=-1 ]; then echo "c.NotebookApp.token='$TOKEN'" >> $HOME/.jupyter/jupyter_notebook_config.py; fi
+
+ENTRYPOINT ["jupyter", "notebook","--ip=0.0.0.0","--allow-root","--no-browser"]
```
Comment (aecio) [inline Dockerfile:30]: Check if `--allow-root` is really necessary, otherwise, it might be better to remove it.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+# Create a user, since we don't want to run as root
+RUN useradd -m eden
+ENV HOME=/home/eden
+WORKDIR $HOME
+USER eden
+COPY --chown=eden examples /home/eden/examples
+
+ARG TOKEN=-1
+RUN mkdir -p $HOME/.jupyter/
+RUN if [ $TOKEN!=-1 ]; then echo "c.NotebookApp.token='$TOKEN'" >> $HOME/.jupyter/jupyter_notebook_config.py; fi
```
Comment (aecio) [inline Dockerfile:28]: Is this token configuration a requirement for JupyterHub? It looks like the token can also be controlled by the environment variable `JUPYTER_TOKEN` (https://jupyter-notebook.readthedocs.io/en/stable/config.html). If possible, using an environment variable may be better as it doesn't require rebuilding the image to change the token.

**Review by aecio — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 2 above)
Comment (EdenWuyifan) [reply Dockerfile:28]: This is used solely for docker build, the builder can add their own token, or simply just without it, like this:
`docker build --build-arg TOKEN='' --build-arg BUILD_OPTION='full' .`

Comment (aecio) [reply Dockerfile:28]: I get that. What I'm not sure of is if there are any advantages of hardcoding the token at build time (as opposed to providing the token at runtime using an environment variable). If you just remove those lines, we could run the image using the following command:

```
docker run -p 8888:8888 -e JUPYTER_TOKEN="<my-token>" ghcr.io/vida-nyu/alpha-automl
```

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    jupyterlab-server==2.16.0
+
+# Install AlphaD3M and dependencies
+# RUN pip3 --disable-pip-version-check install --no-cache-dir \
+# 	alpha-automl
+
+ADD . /alpha-automl/
+WORKDIR /alpha-automl/
+ARG BUILD_OPTION='full'
+RUN pip3 install -e .[$BUILD_OPTION]
```
Comment (aecio) [inline Dockerfile:17]: This build does not work if we pass an empty string or do not pass the `BUILD_OPTION` parameter, i.e., we can't build an image of the default lightweight version. Maybe we should check if the parameter was set, and if not, just run only `pip3 install -e .`

File: kubernetes/values.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      enabled: false
+singleuser:
+  storage:
+    capacity: 10Gi
+  networkPolicy:
+    enabled: false
+  cloudMetadata:
+    blockWithIptables: false
+  image:
+    name: edenwu/alpha_automl
```
Comment (aecio) [inline kubernetes/values.yaml:32]: Now that an initial image was pushed, we can modify this to point to the image at `ghcr.io/vida-nyu/alpha-automl`.

**Review by aecio — APPROVED** [review]
Thanks, @EdenWuyifan!

After these changes, we can run the service as follows.

To let Jupyter auto-generate a token:
```
docker run -p 8888:8888 ghcr.io/vida-nyu/alpha-automl
```

To run using a custom security token:
```
docker run -p 8888:8888 -e JUPYTER_TOKEN="<my-token>" ghcr.io/vida-nyu/alpha-automl
```

If the service is running in a secure environment, the authentication can be disabled:
```
docker run -p 8888:8888 ghcr.io/vida-nyu/alpha-automl --NotebookApp.token=''
```

**Comment by aecio** [comment]
I think we are still missing two things (but we can do it later in a separate PR):
- A README.md file for the `kubernetes` folder with short documentation on how to deploy the Jupyter Hub service;
- A Dockerfile for the `0.3.0.dev0-jupyterhub` version of the image, although I'm not sure the best way to add this without having two near-duplicate files.

**Comment by EdenWuyifan** [comment]
> Thanks, @EdenWuyifan!
> 
> After these changes, we can run the service as follows.
> 
> To let Jupyter auto-generate a token:
> 
> ```
> docker run -p 8888:8888 ghcr.io/vida-nyu/alpha-automl
> ```
> 
> To run using a custom security token:
> 
> ```
> docker run -p 8888:8888 -e JUPYTER_TOKEN="<my-token>" ghcr.io/vida-nyu/alpha-automl
> ```
> 
> If the service is running in a secure environment, the authentication can be disabled:
> 
> ```
> docker run -p 8888:8888 ghcr.io/vida-nyu/alpha-automl --NotebookApp.token=''
> ```

Also, we can use a build option (full, timeseries, nlp...) upon docker build for different building plans:
`docker build --build-arg BUILD_OPTION=full .`
or simply a base version using:
`docker build .`

**Pull request merged**